ons-spark/spark-concepts/partitions.md

```diff
@@ -0,0 +1,941 @@
+````
+## How should I partition my data?
+
+The short answer is- don't worry about it too much!
+
+Let's flip the question around. What are the consequences of getting the number of partitions "wrong"? Most of the time- nothing. You could spend hours trying to optimise the number of partitions and gain a $x2$ increase in processing speed, but it's likely there are other things you can do that will gain $x10$ or $x100$ increase in processing speed. 
+
+As the computer scientist Donald Knuth once wrote, 
+
+>Premature optimization is the root of all evil (or at least most of it) in programming
+>
+>    -> *Computer Programming as an Art (1974)*
+
+More importantly it's good practice to be aware of the size of the DataFrame and the number of partitions as this will help to avoid obvious issues or over-partitioning, under-partitioning and highly skewed processes.
+
+A more accurate answer depends on a variety of factors including: the size of the data, data types, distributions within the data, type of processing and other properties defined in the `SparkSession`. Here are some tips for quick wins:
+
+- as suggested above, the first step is getting the code right. Only look to optimise if you're running into performance issues
+- decreasing the `spark.sql.shuffle.partitions` parameter is sensible for smaller datasets
+- if you google it, you might find the optimum size for a Spark partitions is 100-200MB. In practice, this is impossible to keep track of in an analysis script, but it gives you an idea of what is considered big or small
+- it makes sense for the number of partitions to be some multiple of the number of cores in the Spark session. This will help to avoid redundant cores
+
+Remember, it's only worth experimenting on the *optimum* number if you have a real performance issue. 
+
+## Further resources
+
+PySpark:
+
+- [`.rdd.getNumPartitions()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.getNumPartitions.html)
+- [`.rdd.mapPartitions()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.mapPartitions.html?highlight=mappartitions)
+- [`.coalesce()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.coalesce.html?highlight=coalesce#pyspark.sql.DataFrame.coalesce)
+- [`.repartition()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.repartition.html?highlight=repartition#pyspark.sql.DataFrame.repartition)
+- [`Window.partitionBy()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.Window.partitionBy.html?highlight=partitionby#pyspark.sql.Window.partitionBy)
+- [`DataFrameWriter.partitionBy()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrameWriter.partitionBy.html?highlight=partitionby#pyspark.sql.DataFrameWriter.partitionBy)
+
+sparklyr:
+
+- [`sdf_num_partitions()`](https://www.rdocumentation.org/packages/sparklyr/versions/0.8.2/topics/sdf_num_partitions)
```

Comment:
Links go to rdocumentation rather than spark rstudio; don't worry about this for now as I'm going to sort them all in one go soon